samcli/cli/cli_config_file.py

```diff
@@ -0,0 +1,152 @@
+
+
+def configuration_callback(cmd_name, option_name, identifier_name, saved_callback, provider, ctx, param, value):
+    """
+    Callback for reading the config file.
+
+    Also takes care of calling user specified custom callback afterwards.
+
+    :param cmd_name: `sam` command name derived from click.
+    :param option_name: The name of the option. This is used for error messages.
+    :param identifier_name: `top` level section within configuration file
+    :param saved_callback: User-specified callback to be called later.
+    :param provider:  A callable that parses the configuration file and returns a dictionary
+        of the configuration parameters. Will be called as
+        `provider(file_path, identifier, cmd_name)`.
+    :param ctx: Click context
+    :param param: Click parameter
+    :param value: Specified value for identifier
+    :returns specified callback or the specified value for identifier.
+    """
+    ctx.default_map = ctx.default_map or {}
+    cmd_name = cmd_name or ctx.info_name
+
+    config_file = os.getenv("SAM_CONFIG", os.path.abspath(os.path.join(".", DEFAULT_CONFIG_FILE_NAME)))
+
+    param.default = DEFAULT_IDENTIFER
+    identifier_name = value or identifier_name
+
+    if os.path.isfile(config_file):
+        try:
+            LOG.debug("Config file location: %s", os.path.abspath(config_file))
+            config = provider(config_file, identifier_name, cmd_name)
+
+        except Exception as ex:
+
+            raise click.BadOptionUsage(
+                option_name,
+                "Error reading configuration file: {} with identifier {}, {}".format(
+                    config_file, identifier_name, str(ex)
+                ),
+                ctx,
+            )
+        ctx.default_map.update(config)
+
+    return saved_callback(ctx, param, value) if saved_callback else value
+
+
+def configuration_option(*param_decls, **attrs):
+    """
+    Adds configuration file support to a click application.
+
+    This will create an option of type `STRING` expecting the identifier in the
+    configuration file, by default this identifier is `default`. When specified,
+    the requisite portion of the configuration file is considered as the
+    source of truth.
+
+    The default name of the option is `--identifier`.
+
+    This decorator accepts the same arguments as `click.option`.
+    In addition, the following keyword arguments are available:
+    :param cmd_name: The command name. Default: `ctx.info_name`
+    :param identifier_name: The identifier name. This is used to determine which part of the configuration
+        needs to be read.
+    :param provider:         A callable that parses the configuration file and returns a dictionary
+        of the configuration parameters. Will be called as
+        `provider(file_path, identifier, cmd_name)
+    """
+    param_decls = param_decls or ("--identifier",)
+    option_name = param_decls[0]
+
+    def decorator(f):
```

Comment:
I don't understand the Click API a whole lot. So if you want me to dig into those and review this method in detail, let me know.

Reply:
We are essentially creating a hidden option, whose callback can help populate defaults for that command.